### Import required dependencies

In [1]:
import pandas as pd
import os

## Deliverable 1: Collect the Data

To collect the data that you’ll need, complete the following steps:

1. Using the Pandas `read_csv` function and the `os` module, import the data from the `new_full_student_data.csv` file, and create a DataFrame called student_df. 

2. Use the head function to confirm that Pandas properly imported the data.


In [2]:
# Create the path and import the data
full_student_data = os.path.join('../Resources/new_full_student_data.csv')
student_df = pd.read_csv(full_student_data)

In [3]:
# Verify that the data was properly imported
student_df.head()

student_id      student_name grade           school_name  reading_score  \
0   103880842     Travis Martin   9th  Sullivan High School           59.0   
1    45069750     Michael Brown   9th     Dixon High School           94.7   
2    45024902   Gabriela Lucero   9th    Wagner High School           89.0   
3    62582498  Susan Richardson   9th     Silva High School           69.7   
4    16437227      Sherry Davis  11th    Bowers High School            NaN   

   math_score school_type  school_budget  
0        88.2      Public         961125  
1        73.5     Charter         870334  
2        70.4      Public         846745  
3        80.3      Public         991918  
4        27.5      Public         848324

## Deliverable 2: Prepare the Data

To prepare and clean your data for analysis, complete the following steps:
    
1. Check for and remove all rows with `NaN`, or missing, values in the student DataFrame. 

2. Check for and remove all duplicate rows in the student DataFrame.

3. Use the `str.replace` function to remove the "th" from the grade levels in the grade column.

4. Check data types using the `dtypes` property.

5. Remove the "th" suffix from every value in the grade column using `str` and `replace`.

6. Change the grade colum to the `int` type and verify column types.

7. Use the head (and/or the tail) function to preview the DataFrame.

In [4]:
# Check for null values
student_df.isnull()

student_id  student_name  grade  school_name  reading_score  \
0           False         False  False        False          False   
1           False         False  False        False          False   
2           False         False  False        False          False   
3           False         False  False        False          False   
4           False         False  False        False           True   
...           ...           ...    ...          ...            ...   
19509       False         False  False        False          False   
19510       False         False  False        False          False   
19511       False         False  False        False          False   
19512       False         False  False        False          False   
19513       False         False  False        False          False   

       math_score  school_type  school_budget  
0           False        False          False  
1           False        False          False  
2           False        False          False  
3           False        False          False  
4           False        False          False  
...           ...          ...            ...  
19509       False        False          False  
19510       False        False          False  
19511       False        False          False  
19512       False        False          False  
19513       False        False          False  

[19514 rows x 8 columns]

In [5]:
# Drop rows with null values and verify removal
student_df.isnull().sum()


student_id          0
student_name        0
grade               0
school_name         0
reading_score    1968
math_score        982
school_type         0
school_budget       0
dtype: int64

In [6]:
student_df.dropna()

student_id      student_name grade             school_name  \
0       103880842     Travis Martin   9th    Sullivan High School   
1        45069750     Michael Brown   9th       Dixon High School   
2        45024902   Gabriela Lucero   9th      Wagner High School   
3        62582498  Susan Richardson   9th       Silva High School   
5        74579444   Cynthia Johnson   9th  Montgomery High School   
...           ...               ...   ...                     ...   
19509   109236636     Robert Sawyer  12th       Silva High School   
19510    63239258      David Herman   9th       Woods High School   
19511    95516554        Megan Gill  11th      Wagner High School   
19512    65050383        Lori Stone  11th      Bowers High School   
19513    34720657       Anna Jensen  12th  Montgomery High School   

       reading_score  math_score school_type  school_budget  
0               59.0        88.2      Public         961125  
1               94.7        73.5     Charter         870334  
2               89.0        70.4      Public         846745  
3               69.7        80.3      Public         991918  
5               63.5        76.9     Charter         893368  
...              ...         ...         ...            ...  
19509           43.3        27.2      Public         991918  
19510           52.1        80.4      Public         912243  
19511           93.9        84.1      Public         846745  
19512           94.6        70.9      Public         848324  
19513           82.3        42.4     Charter         893368  

[16667 rows x 8 columns]

In [7]:
# Check for duplicated rows
student_df.duplicated()


0        False
1        False
2        False
3        False
4        False
         ...  
19509    False
19510     True
19511    False
19512    False
19513    False
Length: 19514, dtype: bool

In [8]:
student_df.duplicated().sum()

2168

In [9]:
# Drop duplicated rows and verify removal
student_df = student_df.drop_duplicates()
student_df


student_id      student_name grade             school_name  \
0       103880842     Travis Martin   9th    Sullivan High School   
1        45069750     Michael Brown   9th       Dixon High School   
2        45024902   Gabriela Lucero   9th      Wagner High School   
3        62582498  Susan Richardson   9th       Silva High School   
4        16437227      Sherry Davis  11th      Bowers High School   
...           ...               ...   ...                     ...   
19508    83985333   Deborah Sanders  10th       Silva High School   
19509   109236636     Robert Sawyer  12th       Silva High School   
19511    95516554        Megan Gill  11th      Wagner High School   
19512    65050383        Lori Stone  11th      Bowers High School   
19513    34720657       Anna Jensen  12th  Montgomery High School   

       reading_score  math_score school_type  school_budget  
0               59.0        88.2      Public         961125  
1               94.7        73.5     Charter         870334  
2               89.0        70.4      Public         846745  
3               69.7        80.3      Public         991918  
4                NaN        27.5      Public         848324  
...              ...         ...         ...            ...  
19508           60.5        64.6      Public         991918  
19509           43.3        27.2      Public         991918  
19511           93.9        84.1      Public         846745  
19512           94.6        70.9      Public         848324  
19513           82.3        42.4     Charter         893368  

[17346 rows x 8 columns]

In [10]:
# Check data types using dtypes
student_df.dtypes

student_id         int64
student_name      object
grade             object
school_name       object
reading_score    float64
math_score       float64
school_type       object
school_budget      int64
dtype: object

In [11]:
# Examine the grade column to understand why it is not an int
student_df["grade"]

0         9th
1         9th
2         9th
3         9th
4        11th
         ... 
19508    10th
19509    12th
19511    11th
19512    11th
19513    12th
Name: grade, Length: 17346, dtype: object

In [12]:
# Remove the non-numeric characters and verify the contents of the column, using str and replace
student_df['grade'] = student_df['grade'].str.replace('th', '')
student_df['grade']

0         9
1         9
2         9
3         9
4        11
         ..
19508    10
19509    12
19511    11
19512    11
19513    12
Name: grade, Length: 17346, dtype: object

In [13]:
# Change the grade column to the int type and verify column types
student_df['grade'] = student_df['grade'].astype(int)
student_df.dtypes

student_id         int64
student_name      object
grade              int64
school_name       object
reading_score    float64
math_score       float64
school_type       object
school_budget      int64
dtype: object

## Deliverable 3: Summarize the Data

Describe the data using summary statistics on the data as a whole and on individual columns.

1. Generate the summary statistics for each DataFrame by using the `describe` function.

2. Display the mean math score using the `mean` function. 

2. Store the minimum reading score as `min_reading_score`.

In [14]:
# Display summary statistics for the DataFrame
student_df.describe()

student_id         grade  reading_score    math_score  school_budget
count  1.734600e+04  17346.000000   15611.000000  16479.000000   17346.000000
mean   6.971921e+07     10.356393      72.371046     64.715899  893753.090741
std    3.458822e+07      1.097201      15.206900     15.827649   53950.325200
min    1.000906e+07      9.000000      10.500000      3.700000  817615.000000
25%    3.962676e+07      9.000000      62.200000     54.500000  846745.000000
50%    6.961437e+07     10.000000      73.800000     65.300000  893368.000000
75%    9.944595e+07     11.000000      84.000000     76.000000  956438.000000
max    1.299997e+08     12.000000     100.000000    100.000000  991918.000000

In [15]:
# Display the mean "math_score" using the mean function
student_df['math_score'].mean()

64.71589902299891

In [16]:
# Store the minimum "reading_score"column as min_reading_score
student_df['reading_score'].min()

10.5

## Deliverable 4: Drill Down into the Data

Drill down to specific rows, columns, and subsets of the data.

To drill down into the data, complete the following steps:

1. Use `loc` to display the grade column.

2. Use `iloc` to display the first 3 rows and columns 3, 4, and 5.

3. Show the rows for grade nine using `loc`.

4. Store the row with the minimum overall reading score as `min_reading_row` using `loc` and the `min_reading_score` found in Deliverable 3.

5. Find the reading scores for the school and grade from the output of step three using `loc` with multiple conditional statements.

6. Using conditional statements and `loc` or `iloc`, find the mean reading score for all students in grades 11 and 12 combined.

In [17]:
# Use loc to display the grade column
student_df['grade']

0         9
1         9
2         9
3         9
4        11
         ..
19508    10
19509    12
19511    11
19512    11
19513    12
Name: grade, Length: 17346, dtype: int64

In [18]:
# Use `iloc` to display the first 3 rows and columns 3, 4, and 5.
student_df.iloc[0:3, 3:6] 

school_name  reading_score  math_score
0  Sullivan High School           59.0        88.2
1     Dixon High School           94.7        73.5
2    Wagner High School           89.0        70.4

In [19]:
# Select the rows for Grade 9, and display their summary stats by using loc and describe
student_df.loc[student_df['grade'] == 9].describe()




student_id   grade  reading_score   math_score  school_budget
count  4.832000e+03  4832.0    4353.000000  4587.000000    4832.000000
mean   6.973426e+07     9.0      69.322927    66.686222  898548.483858
std    3.468975e+07     0.0      15.287125    16.692797   54770.293534
min    1.000906e+07     9.0      17.900000     5.300000  817615.000000
25%    3.957139e+07     9.0      59.100000    56.100000  846745.000000
50%    6.984037e+07     9.0      70.100000    67.800000  893368.000000
75%    9.942976e+07     9.0      80.600000    78.600000  957299.000000
max    1.299997e+08     9.0      99.900000   100.000000  991918.000000

In [20]:
# Store the row with the minimum overall reading score in "min_reading_row" by using loc and the "min_reading_score" variable from Deliverable 3
min_reading_score = student_df['reading_score'].min()
min_reading_row = student_df.loc[student_df["reading_score"] == min_reading_score]
min_reading_row



student_id    student_name  grade        school_name  reading_score  \
3706    81758630  Matthew Thomas     10  Dixon High School           10.5   

      math_score school_type  school_budget  
3706        58.4     Charter         870334

In [21]:
# Use loc with conditionals to select all reading scores from 10th graders at Dixon High School.
student_df.loc[(student_df['school_name'] == "Dixon High School") & (student_df['grade'] == 10), ['school_name', 'reading_score']]

school_name  reading_score
45     Dixon High School           71.1
60     Dixon High School           59.5
69     Dixon High School           88.6
94     Dixon High School           81.5
100    Dixon High School           95.3
...                  ...            ...
19283  Dixon High School           52.9
19306  Dixon High School           58.0
19344  Dixon High School           38.0
19368  Dixon High School           84.4
19445  Dixon High School           43.9

[654 rows x 2 columns]

In [22]:
# Find the mean reading score for all students in grades 11 and 12 combined.
student_df.loc[student_df['grade'] >= 11].mean

<bound method NDFrame._add_numeric_operations.<locals>.mean of        student_id       student_name  grade             school_name  \
4        16437227       Sherry Davis     11      Bowers High School   
6        52067716      Thomas Wilson     12  Montgomery High School   
9        76307038     Kristen Torres     12      Campos High School   
13       60241809  Daniel Washington     11      Wagner High School   
15       19432614      Richard Simon     11       Green High School   
...           ...                ...    ...                     ...   
19507    42009506     Peter Williams     12      Bowers High School   
19509   109236636      Robert Sawyer     12       Silva High School   
19511    95516554         Megan Gill     11      Wagner High School   
19512    65050383         Lori Stone     11      Bowers High School   
19513    34720657        Anna Jensen     12  Montgomery High School   

       reading_score  math_score school_type  school_budget  
4                NaN  

## Deliverable 5: Make Comparisons Between District and Charter Schools

Compare district vs charter schools for budget, size, and scores.

Make comparisons within your data by completing the following steps:

1. Using the `groupby` and `mean` functions, look at the average reading and math scores per school type.

1. Using the `groupby` and `count` functions, find the total number of students at each school.

2. Using the `groupby` and `mean` functions, find the average budget per grade for each school type.

In [24]:
# Use groupby and mean to find the average reading and math scores for each school type.

avg_by_school = student_df.groupby(['school_name', 'reading_score', 'math_score']).mean()
avg_by_school.loc[:, ["reading_score", "math_score"]]


KeyError: "None of [Index(['reading_score', 'math_score'], dtype='object')] are in the [columns]"

In [ ]:
# The average budget for each school type is displayed

In [ ]:
# Use the `groupby`, `count`, and `sort_values` functions to find the
# total number of students at each school and sort from most students to least students.


# Deliverable 6: Summarize Your Findings
In the cell below, write a few sentences to describe any discoveries you made while performing your analysis along with any additional analysis you believe would be worthwhile.

*The data analyzed includes student names, reading and math scores, and budgets of public and charter schools.* 

* The mean for math scores in all schools is 64.7 and mean for reading is 72.3. 
* All schools received a budget between $817,000 to $992,000. 

*Further analysis could: 
    -pull the top three scoring schools vs the lower 3 scoring schools
    -correlate the school budgets with reading/math scores. 
    -correlate the school budgets with the type of school. Do public or charter schools get greater funding? 
    